WPThemeReview/Sniffs/Templates/RequiredFunctionSniff.php

```diff
@@ -0,0 +1,231 @@
+			$nextPtrCode    = $tokens[ $nextPtr ]['code'];
+
+			// Check for attribute not allowed.
+			if (
+				false === $foundAttribute &&
+				in_array( $nextPtrCode, Tokens::$textStringTokens, true ) &&
+				false !== stripos( $nextPtrContent, $tagAttr . '=' )
+			) {
+				$foundAttribute = true;
+			}
+
+			// Check for required function call.
+			if (
+				false === $foundFunction &&
+				in_array( $nextPtrCode, Tokens::$functionNameTokens, true ) &&
+				false !== strpos( $nextPtrContent, $tagFn )
+			) {
+
+				// Check next non-whitespace token for opening parens.
+				$next = $phpcsFile->findNext( Tokens::$emptyTokens, ( $nextPtr + 1 ), null, true );
+
+				if ( ! $next || ! isset( $tokens[ $next ] ) ) {
+					break; // Nothing left.
+				}
+
+				// Verify function( $param = 'optional' ) type.
+				if ( 'PHPCS_T_OPEN_PARENTHESIS' === $tokens[ $next ]['code'] ) {
+
+					// Skip over contents to closing parens in stack.
+					if ( isset( $tokens[ $next ]['parenthesis_closer'] ) ) {
+						$nextPtr       = $tokens[ $next ]['parenthesis_closer'];
+						$foundFunction = true;
+					}
+				}
+
+				continue;
+			}
+
+			// Check for searched tag matched closing bracket.
+			if (
+				in_array( $nextPtrCode, Tokens::$textStringTokens, true ) &&
+				'>' === substr( $nextPtrContent, -1 )
+			) {
+				$this->tag[ $filename ] = false;
+				$foundEnd               = true;
+				break;
+			}
+
+			// Increment stack to next non-whitespace token.
+			$next = $phpcsFile->findNext( Tokens::$emptyTokens, ( $nextPtr + 1 ), null, true );
+
+			if ( ! $next || ! isset( $tokens[ $next ] ) ) {
+				break; // Short circuit loop as there's not anything left.
+			}
+
+			$nextPtr = $next;
+
+		} while ( false === $foundEnd ); // Loop until matched closing bracket is found for searched tag.
+
+		// Required function not found.
+		if ( false === $foundFunction ) {
+			$phpcsFile->addError(
+				"Themes must call {$tagFn}() inside <{$tagName}> tags.",
+				$stackPtr,
+				"RequiredFunction{$pascal}"
+			);
+
+			return;
+		}
+
+		// Atrribute is not allowed.
+		if ( true === $foundAttribute ) {
+			$phpcsFile->addError(
+				"Attribute '{$tagAttr}' is not allowed on <{$tagName}> tags. Themes must call {$tagFn}() instead.",
+				$stackPtr,
+				"DisallowedAttribute{$pascal}"
+			);
+
+			return;
+		}
+	}
+
+	/**
+	 * Cleans string for parsing.
+	 *
+	 * This cleans whitespace chars and single/double quotes
+	 * from string.  Primary used to check T_CONSTANT_ENCAPSED_STRING
+	 * and T_DOUBLE_QUOTED_STRING for closing HTML brackets.  This is
+	 * because < and > are valid attribute values, and a strpos wouldn't
+	 * be enough.
+	 *
+	 * Strips:
+	 * ' '   : Whitespace
+	 * '"'   : double quote
+	 * '''   : single quote
+	 * '\t'  : tab
+	 * '\n'  : newline
+	 * '\r'  : carriage return
+	 * '\0'  : NUL-byte
+	 * '\x0B': vertical tab
+	 *
+	 * @param string $str String to clean.
+	 *
+	 * @return string Cleaned string.
+	 */
+	private function clean_str( $str ) {
```

Comment:
If this could be used in other sniffs, we could do what @jrfnl suggested in the #213 PR about `strip_quotes` method - create a `WPThemeReview\Utils\Common` class with these kind of utilities 🙂

Reply:
Funny you mentioned it - I initially made a Util class for this while testing it out, but after looking over WPThemeReview I didn't see one anywhere else, or a common area, so I just figured the "norm" was to keep functions self contained per class.  I noticed some WPCS stuff that would be useful, but I haven't worked my way to that level lol.

It looks like we are all trying to address the same issues revolving around `T_CONSTANT_ENCAPSED_STRING` and `T_DOUBLE_QUOTED_STRING` though.  I don't recall all of the specifics, but I think I needed to trim single quotes in addition to the double quotes added for ensuring the functions were being caught inside of the actual HTML tags where attributes are added, not inside of any quoted areas if it's broken up between HTML/PHP, and accounting for an attribute containing a `>` character.  Not very common but does happen sometimes.  Something like this scenario - if I remember correctly (I could be wrong):

```
echo '<body id="' . $something . '"' . body_class() . "data-title='Some > Title'" . '>';
```

There were some other weird scenarios with doing things like mixing heredoc/functions(print,
 echo, printf etc)/HTML together which doesn't work out perfectly all of the time, but I also think if someone is doing something that far out of the ordinary they obviously have no coding standards or desire to follow any.  I thought of a solution afterwards, but it would require doing some weird tracking of the opening/closing quotes, which seemed like a waste of time to cover those fringe cases ( realistically probably 0.000000000001%/never encountering it ).

I'm not sure if this function will be useful anywhere else though, but I could see just the `strip_quotes` being used commonly.  I think a Util class would be nice to have though, and there might be more things as we expand the sniffs further where we realize overlap in some functionality.  I honestly have no clue since this is the first time writing a phpcs sniff :laughing:

Reply:
> ```php
> echo '<body id="' . $something . '"' . body_class() . "data-title='Some > Title'" . '>';
> ```

That would break the HTML. The `>` in `Some > Title` needs to be encoded as `&gt;` for this to be valid HTML.

Reply:
Re: `Util` class or something along those lines.

I think this would be a good idea.

If I remember correctly, the reasons why in WPCS all utility methods are in the abstract `Sniff` class and all classes extend it, is historic.
PHPCS 2.x file auto-loading didn't support helper files as easily, but the PHPCS 3.x file auto-loading is more flexible.

As far as I can see, as long as the file is within the standards directory, is namespaced based on the path and the file name doesn't end on `Sniff`, this should work in PHPCS 3.x out of the box.

A `use` statement for the class in sniffs using it should then be able to take care of the rest.

Maybe I suggest a `WPThemeReview\Helpers\StringUtils.php` file containing a class named `StringUtils` which would then contain the (`static`) functions ?
This will allow for adding more `Util` type classes later for functions addressing other language structures, keeping it all modular instead of having one massive `Util` class.

Alternatively, for the `strip_quotes()` method from WPCS, I'd happily accept a PR to make that a `static` method in WPCS.
For the time being, that should allow you to use it even when a sniff doesn't extend the WPCS `Sniff` class (to bridge the time until the method is available in PHPCS).